diff --git a/build.sh b/build.sh
index ae8607b..89a182b 100755
--- a/build.sh
+++ b/build.sh
@@ -343,6 +343,12 @@ case $MACH in
             # Distro to poky-swi-ext to change SDKPATH
             DISTRO="poky-swi-ext"
         fi
+        if [ $QEMU = true ]; then
+            bitbake-layers add-layer $SWI/meta-$MACH-virt
+            if [ $? -ne 0 ];then
+                bitbake-layers add-layer $SWI/meta-swi-sdxprairie-virt
+            fi
+        fi
         ;;
 
 esac
diff --git a/repo.mk b/repo.mk
index f290cbd..848858b 100644
--- a/repo.mk
+++ b/repo.mk
@@ -324,10 +324,17 @@ ifeq ($(QEMU),1)
   QEMU_ARGS := -Q
 endif
 
+ifneq ($(BUILDDIR),)
+COMMON_SRC := \
+				$(COMMON_MACH) \
+				${QEMU_ARGS} \
+				-b $(BUILDDIR)
+else
 COMMON_SRC := \
 				$(COMMON_MACH) \
 				${QEMU_ARGS} \
 				-b build_src
+endif
 
 ifeq ($(PROPRIETARY_BUILD),1)
   COMMON_SRC += -w $(APPS_DIR) $(FW_VERSION_ARG) -s
